src/OrleansServiceBus/Providers/EventDataGeneratorStreamProvider/IEventDataGenerator.cs

```diff
@@ -0,0 +1,178 @@
+            {
+                //finish producing drain events first
+                return this.drainEventCount > maxCount ? maxCount : this.drainEventCount;
+            }
+            else
+            {
+                //after producing drain events, mimic real events producing pattern : random count
+                return this.seed.Next(maxCount + 1);
+            }
+        }
+
+        IEnumerable<int> GenerateEvent(int sequenceNumber)
+        {
+            var events = new List<int>();
+            events.Add(sequenceNumber);
+            return events;
+        }
+    }
+
+    /// <summary>
+    /// EHPartitionDataGenerator generate data for a EH partition, which can include data from different streams
+    /// </summary>
+    internal class EventHubPartitionDataGenerator : IDataGenerator<EventData>, IStreamDataGeneratingController
+    {
+        //differnt stream in the same partition should use the same sequenceNumberCounter
+        private IntCounter sequenceNumberCounter = new IntCounter();
+        private Logger logger;
+        private List<IStreamDataGenerator<EventData>> generators;
+        private SerializationManager serializationManager;
+        private EventHubGeneratorStreamProviderSettings settings;
+        public EventHubPartitionDataGenerator(Logger logger, SerializationManager serializationManager, EventHubGeneratorStreamProviderSettings settings)
+        {
+            this.logger = logger.GetSubLogger(this.GetType().Name);
+            this.generators = new List<IStreamDataGenerator<EventData>>();
+            this.serializationManager = serializationManager;
+            this.settings = settings;
+        }
+        public void AddDataGeneratorForStream(IStreamIdentity streamId)
+        {
+            var generator = (IStreamDataGenerator<EventData>)Activator.CreateInstance(settings.StreamDataGeneratorType,
+                streamId, settings, this.logger, this.serializationManager);
+            generator.SequenceNumberCounter = sequenceNumberCounter;
+            this.logger.Info($"Data generator set up on stream {streamId.Namespace}-{streamId.Guid.ToString()}.");
+            this.generators.Add(generator);
+        }
+        public void StopProducingOnStream(IStreamIdentity streamId)
+        {
+            this.generators.ForEach(generator => {
+                if (generator.StreamId.Equals(streamId))
+                {
+                    generator.ShouldProduce = false;
+                    this.logger.Info($"Stop producing data on stream {streamId.Namespace}-{streamId.Guid.ToString()}.");
+                }
+            });
+        }
+        public bool TryReadEvents(int maxCount, TimeSpan waitTime, out IEnumerable<EventData> events)
+        {
+            var eventDataList = new List<EventData>();
+            this.generators.ForEach(generator => {
+                IEnumerable<EventData> eventData;
+                if (generator.TryReadEvents(maxCount, waitTime, out eventData))
```

Comment:
won't this read maxcount * generater.count events rather than just maxcount events?